### PR TITLE
Doc updates related to TLSv1.3 PSK early data

### PR DESCRIPTION
**Pull request by raja-ashok**
- 498e961fb90ac825961a168412a8303d42cac5e7 For TLSv1.3 out-of-band PSK handshake recommendation is to use new callback (`SSL_psk_use_session_cb_func` and `SSL_psk_find_session_cb_func`) but older callback (`SSL_psk_client_cb_func` and `SSL_psk_server_cb_func`) also gets supported with AES_128_GCM ciphersuite. But TLSv1.3 early data is not allowed to exchange with older callback usage. This limiation of older callback is updated in doc.
- 4342cb75fd7512f247f31ce5b004e170f3dd9f1c Updated the doc as early data can be exchanged while resuming previous session or during out-of-band PSK session.

Needs backport to `1.1.1` also.

##### Checklist

- [x] documentation is added or updated

### CODE REVIEW

File: doc/man3/SSL_CTX_set_psk_client_callback.pod [hunk 1]
```diff
@@ -123,7 +123,9 @@ and it will use that in preference. If no such callback is present then it will
 check to see if a callback has been set via SSL_CTX_set_psk_client_callback() or
 SSL_set_psk_client_callback() and use that. In this case the B<hint> value will
 always be NULL and the handshake digest will default to SHA-256 for any returned
-PSK.
+PSK. TLSv1.3 early data exchange is possible in PSK connections only with
+B<SSL_psk_use_session_cb_func> callback, not possible with
+B<SSL_psk_client_cb_func> callback.
```
Comment (kaduk) [inline doc/man3/SSL_CTX_set_psk_client_callback.pod:128]: A couple nits here: "only with the B<SSL_psk_use_session_cb_func> callback, and are not possible with the B<SSL_psk_client_cb_func> callback"

File: doc/man3/SSL_CTX_use_psk_identity_hint.pod [hunk 2]
```diff
@@ -83,7 +83,9 @@ via SSL_CTX_set_psk_find_session_callback() or SSL_set_psk_find_session_callback
 and it will use that in preference. If no such callback is present then it will
 check to see if a callback has been set via SSL_CTX_set_psk_server_callback() or
 SSL_set_psk_server_callback() and use that. In this case the handshake digest
-will default to SHA-256 for any returned PSK.
+will default to SHA-256 for any returned PSK. TLSv1.3 early data exchange is
+possible in PSK connections only with B<SSL_psk_find_session_cb_func> callback,
+not possible with B<SSL_psk_server_cb_func> callback.
```
Comment (kaduk) [inline doc/man3/SSL_CTX_use_psk_identity_hint.pod:88]: (same here)

File: doc/man3/SSL_read_early_data.pod [hunk 3]
```diff
@@ -58,10 +58,11 @@ SSL_set_allow_early_data_cb
 These functions are used to send and receive early data where TLSv1.3 has been
 negotiated. Early data can be sent by the client immediately after its initial
 ClientHello without having to wait for the server to complete the handshake.
-Early data can only be sent if a session has previously been established with
-the server, and the server is known to support it. Additionally these functions
-can be used to send data from the server to the client when the client has not
-yet completed the authentication stage of the handshake.
+Early data can be sent if a session has previously been established with the
+server or establishing new session using an out-of-band PSK, and the server is
```
Comment (kaduk) [inline doc/man3/SSL_read_early_data.pod:62]:
```suggestion
server or when establishing a new session using an out-of-band PSK, and only when the server is
```

File: doc/man3/SSL_CTX_set_psk_client_callback.pod (hunk 1 above)
Comment (raja-ashok) [reply doc/man3/SSL_CTX_set_psk_client_callback.pod:128]: Then I feel the start of the phrase should be `TLSv1.3 early data exchanges are`.

File: doc/man3/SSL_read_early_data.pod (hunk 3 above)
Comment (raja-ashok) [reply doc/man3/SSL_read_early_data.pod]: Fixed

**Comment by nmaji2020** [comment]
I want these documents to make Gmail with the desired domain skype ...

PKCS#12👉      PfX or P12

File: doc/man3/SSL_CTX_set_psk_client_callback.pod [hunk 4]
```diff
@@ -123,7 +123,9 @@ and it will use that in preference. If no such callback is present then it will
 check to see if a callback has been set via SSL_CTX_set_psk_client_callback() or
 SSL_set_psk_client_callback() and use that. In this case the B<hint> value will
 always be NULL and the handshake digest will default to SHA-256 for any returned
-PSK.
+PSK. TLSv1.3 early data exchanges are possible in PSK connection only with the
```
Comment (kaduk) [inline doc/man3/SSL_CTX_set_psk_client_callback.pod:126]: nit: "PSK connections" plural

File: doc/man3/SSL_CTX_use_psk_identity_hint.pod [hunk 5]
```diff
@@ -83,7 +83,9 @@ via SSL_CTX_set_psk_find_session_callback() or SSL_set_psk_find_session_callback
 and it will use that in preference. If no such callback is present then it will
 check to see if a callback has been set via SSL_CTX_set_psk_server_callback() or
 SSL_set_psk_server_callback() and use that. In this case the handshake digest
-will default to SHA-256 for any returned PSK.
+will default to SHA-256 for any returned PSK. TLSv1.3 early data exchanges are
+possible in PSK connection only with the B<SSL_psk_find_session_cb_func>
```
Comment (kaduk) [inline doc/man3/SSL_CTX_use_psk_identity_hint.pod:87]: (and here)

File: doc/man3/SSL_read_early_data.pod [hunk 6]
```diff
@@ -58,10 +58,11 @@ SSL_set_allow_early_data_cb
 These functions are used to send and receive early data where TLSv1.3 has been
 negotiated. Early data can be sent by the client immediately after its initial
 ClientHello without having to wait for the server to complete the handshake.
-Early data can only be sent if a session has previously been established with
-the server, and the server is known to support it. Additionally these functions
-can be used to send data from the server to the client when the client has not
-yet completed the authentication stage of the handshake.
+Early data can be sent if a session has previously been established with the
+server or when establishing a new session using an out-of-band PSK, and only
+when the server is known to support it. Additionally these functions can be used
```
Comment (kaduk) [inline doc/man3/SSL_read_early_data.pod:63]: I personally would put a comma after "Additionally" but I don't think it's strictly required.

**Comment by kaduk** [comment]
I added labels for master and 1.1.1, since I think it should apply there as well.
Needs OTC review.

File: doc/man3/SSL_CTX_set_psk_client_callback.pod (hunk 4 above)
Comment (raja-ashok) [reply doc/man3/SSL_CTX_set_psk_client_callback.pod]: Fixed.

File: doc/man3/SSL_CTX_use_psk_identity_hint.pod (hunk 5 above)
Comment (raja-ashok) [reply doc/man3/SSL_CTX_use_psk_identity_hint.pod]: Fixed.

File: doc/man3/SSL_read_early_data.pod (hunk 6 above)
Comment (raja-ashok) [reply doc/man3/SSL_read_early_data.pod:63]: I also feel not required.

**Comment by kaduk** [comment]
Reconfirm +1

**Comment by kaduk** [comment]
merged to master and 1.1.1; closing.
Thanks for the fixes!

**Pull request closed without merging**